File: articles/stage-theories-of-development-promise-power-and-critique/c/stage_theory_summary.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LINE 16384

int main(int argc, char *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s data/stage_theory_development_panel.csv\n", argv[0]);
        return 1;
    }

    FILE *fp = fopen(argv[1], "r");
    if (!fp) {
        perror("Unable to open CSV file");
        return 1;
    }

    char line[MAX_LINE];
    if (!fgets(line, sizeof(line), fp)) {
        fprintf(stderr, "CSV is empty\n");
        fclose(fp);
        return 1;
    }

    int score_col = -1, readiness_col = -1, support_col = -1, stress_col = -1, logistic_col = -1, col = 0;
    char *token = strtok(line, ",");

    while (token != NULL) {
        token[strcspn(token, "\r\n")] = 0;
        if (strcmp(token, "development_score") == 0) score_col = col;
        if (strcmp(token, "transition_readiness") == 0) readiness_col = col;
        if (strcmp(token, "current_support") == 0) support_col = col;
        if (strcmp(token, "chronic_stress") == 0) stress_col = col;
        if (strcmp(token, "logistic_transition") == 0) logistic_col = col;
        token = strtok(NULL, ",");
        col++;
    }

    if (score_col < 0 || readiness_col < 0 || support_col < 0 || stress_col < 0 || logistic_col < 0) {
        fprintf(stderr, "Required columns not found\n");
        fclose(fp);
        return 1;
    }

    long count = 0;
    double score_sum = 0.0, readiness_sum = 0.0, support_sum = 0.0, stress_sum = 0.0, logistic_sum = 0.0;

    while (fgets(line, sizeof(line), fp)) {
        int current_col = 0;
        double score = 0.0, readiness = 0.0, support = 0.0, stress = 0.0, logistic = 0.0;
        char *field = strtok(line, ",");

        while (field != NULL) {
            if (current_col == score_col) score = atof(field);
            if (current_col == readiness_col) readiness = atof(field);
            if (current_col == support_col) support = atof(field);
            if (current_col == stress_col) stress = atof(field);
            if (current_col == logistic_col) logistic = atof(field);
            field = strtok(NULL, ",");
            current_col++;
        }

        count++;
        score_sum += score;
        readiness_sum += readiness;
        support_sum += support;
        stress_sum += stress;
        logistic_sum += logistic;
    }

    fclose(fp);

    printf("Rows analyzed: %ld\n", count);
    printf("Mean development_score: %.4f\n", score_sum / count);
    printf("Mean transition_readiness: %.4f\n", readiness_sum / count);
    printf("Mean current_support: %.4f\n", support_sum / count);
    printf("Mean chronic_stress: %.4f\n", stress_sum / count);
    printf("Mean logistic_transition: %.4f\n", logistic_sum / count);
    return 0;
}

File: articles/stage-theories-of-development-promise-power-and-critique/go/stage_theory_summary.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"strconv"
)

func col(header []string, name string) int {
	for i, h := range header {
		if h == name {
			return i
		}
	}
	return -1
}

func main() {
	if len(os.Args) < 2 {
		log.Fatal("Usage: go run stage_theory_summary.go data/stage_theory_development_panel.csv")
	}
	f, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	rows, err := csv.NewReader(f).ReadAll()
	if err != nil {
		log.Fatal(err)
	}
	if len(rows) < 2 {
		log.Fatal("CSV has no data rows")
	}

	header := rows[0]
	scoreIdx := col(header, "development_score")
	readinessIdx := col(header, "transition_readiness")
	supportIdx := col(header, "current_support")
	stressIdx := col(header, "chronic_stress")
	logisticIdx := col(header, "logistic_transition")
	if scoreIdx < 0 || readinessIdx < 0 || supportIdx < 0 || stressIdx < 0 || logisticIdx < 0 {
		log.Fatal("Required columns missing")
	}

	count := 0
	scoreSum, readinessSum, supportSum, stressSum, logisticSum := 0.0, 0.0, 0.0, 0.0, 0.0

	for _, row := range rows[1:] {
		score, e1 := strconv.ParseFloat(row[scoreIdx], 64)
		readiness, e2 := strconv.ParseFloat(row[readinessIdx], 64)
		support, e3 := strconv.ParseFloat(row[supportIdx], 64)
		stress, e4 := strconv.ParseFloat(row[stressIdx], 64)
		logistic, e5 := strconv.ParseFloat(row[logisticIdx], 64)
		if e1 != nil || e2 != nil || e3 != nil || e4 != nil || e5 != nil {
			continue
		}
		count++
		scoreSum += score
		readinessSum += readiness
		supportSum += support
		stressSum += stress
		logisticSum += logistic
	}

	fmt.Printf("Rows analyzed: %d\n", count)
	fmt.Printf("Mean development_score: %.4f\n", scoreSum/float64(count))
	fmt.Printf("Mean transition_readiness: %.4f\n", readinessSum/float64(count))
	fmt.Printf("Mean current_support: %.4f\n", supportSum/float64(count))
	fmt.Printf("Mean chronic_stress: %.4f\n", stressSum/float64(count))
	fmt.Printf("Mean logistic_transition: %.4f\n", logisticSum/float64(count))
}
